compiler/fir/fir2ir/src/org/jetbrains/kotlin/fir/backend/generators/DelegatedMemberGenerator.kt

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.backend.generators
+
+import org.jetbrains.kotlin.descriptors.CallableMemberDescriptor
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.descriptors.isOverridable
+import org.jetbrains.kotlin.fir.backend.*
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.declarations.impl.IrFunctionImpl
+import org.jetbrains.kotlin.ir.declarations.impl.IrPropertyImpl
+import org.jetbrains.kotlin.ir.declarations.impl.IrValueParameterImpl
+import org.jetbrains.kotlin.ir.descriptors.WrappedPropertyDescriptor
+import org.jetbrains.kotlin.ir.descriptors.WrappedSimpleFunctionDescriptor
+import org.jetbrains.kotlin.ir.descriptors.WrappedValueParameterDescriptor
+import org.jetbrains.kotlin.ir.expressions.impl.*
+import org.jetbrains.kotlin.ir.types.classOrNull
+import org.jetbrains.kotlin.ir.types.impl.IrSimpleTypeImpl
+import org.jetbrains.kotlin.ir.types.isNothing
+import org.jetbrains.kotlin.ir.types.isUnit
+import org.jetbrains.kotlin.ir.util.defaultType
+import org.jetbrains.kotlin.ir.util.hasAnnotation
+import org.jetbrains.kotlin.ir.util.isFakeOverride
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.resolve.DescriptorUtils
+
+/**
+ * A generator for delegated members from implementation by delegation.
+ *
+ * It assumes a synthetic field with the super-interface type has been created for the delegate expression. It looks for delegatable
+ * methods and properties in the super-interface, and creates corresponding members in the subclass.
+ * TODO: generic super interface types and generic delegated members.
+ */
+internal class DelegatedMemberGenerator(
+    private val components: Fir2IrComponents
+) : Fir2IrComponents by components {
+
+    // Generate delegated members for [subClass]. The synthetic field [irField] has the super interface type.
+    fun generateDelegatedMembers(irField: IrField, subClass: IrClass) {
```

Comment:
Nit: naming suggestion: `generate` would be good enough, since the call-site would look like `delegateMemberGenerator.generate(...)`

Reply:
done